crates/fluent/src/lib.rs

```diff
@@ -0,0 +1,255 @@
+            Some(message) => {
+                let pattern = message.value().expect("Message has no value.");
+                let value = bundle.format_pattern(pattern, args, &mut errors);
+                // NOTE: Unused arguments are not considered errors.
+                if !errors.is_empty() {
+                    let mut error_message = format!(
+                        "Unexpected formatting errors occurred for message ID \"{id}\" in language \"{}\":\n",
+                        bundle.locales[0].language
+                    );
+                    for error in errors {
+                        error_message.push_str(&format!("{error}\n"));
+                    }
+                    error_message.push_str("\nPlease report this.");
+                    return Err(error_message);
+                }
+                return Ok(value);
+            }
+            None => {
+                continue;
+            }
+        }
+    }
+    Err(format!("Message \"{id}\" not available in any catalog."))
+}
+
+/// Use this to construct `FluentID`s.
+#[macro_export]
+#[cfg(feature = "fluent-extract")]
+macro_rules! fluent_id {
+    ($id:literal) => {
+        $crate::FluentID::new(fish_fluent_extraction::fluent_extract!($id))
+    };
+}
+#[macro_export]
+#[cfg(not(feature = "fluent-extract"))]
+macro_rules! fluent_id {
+    ($id:literal) => {
+        $crate::FluentID::new($id)
+    };
+}
+
+/// Takes a string literal or a [`FluentID`].
+/// If the argument is a string literal, it is converted to a [`FluentID`],
+/// which marks it for extraction.
+#[macro_export]
+macro_rules! ensure_fluent_id {
+    ($id:literal) => {
+        $crate::fluent_id!($id)
+    };
+    ($id:expr) => {
+        $id
+    };
+}
+
+/// Takes a [`FluentID`] and an [`Option<&FluentArgs>`].
+/// Returns a localized string according to the current language precedence,
+/// or panics if an error occurs.
+#[macro_export]
+macro_rules! localize_impl {
+    ($id:expr, $args:expr) => {
+        match $crate::format_localized(&$crate::ensure_fluent_id!($id), $args) {
+            Ok(localization) => localization,
+            Err(e) => {
+                panic!("{e}");
+            }
+        }
+    };
+}
+
+/// Call this to localize a message with Fluent.
+/// The first argument can be a `&FluentID` or a string literal, which will then be converted to a
+/// `&FluentID` implicitly.
+/// Note that Fluent attributes are not supported.
+/// If arguments need to be passed, this can be done by specifying a `&FluentArgs` value as the
+/// second argument, or by adding comma-separated `key: value` pairs.
+#[macro_export]
+macro_rules! localize {
+    // If no args are needed
+    ($id:expr) => {
+        $crate::localize_impl!($id, None)
+    };
+    // For passing pre-built `&FluentArgs`
+    ($id:expr, $args:expr) => {
+        $crate::localize_impl!($id, Some($args))
+    };
+    // For passing args in `key: value` format
+    ($id:expr, $($key:literal: $value:expr),+ $(,)?) => {
```

Comment:
I guess if we replace + with *,
we could maybe get rid of the extra rule for 0 arguments.
Not sure if this makes anything (like compile times) significantly worse,
I don't expect it to.

I guess the difference would be that we pass
Some(vec![]) instead of None to fluent,
but that shouldn't make a difference, right?

Reply:
makes me wonder why Fluent takes an `Option<FluentArgs>` when `FluentArgs` would suffice.. I guess this makes the API easier to call in the common case.